tools/mlir_to_verilog_main.py

```diff
@@ -0,0 +1,78 @@
+import argparse
+import dataclasses
+import logging
+import os
+import sys
+from typing import Dict
+
+from magma.backend.mlir.mlir_to_verilog import mlir_to_verilog, MlirToVerilogOpts
+from magma.common import slice_opts
+
+
+logging.basicConfig(level=os.environ.get("LOGLEVEL", "WARNING").upper())
+
+
+def _field_to_argument_params(field: dataclasses.Field) -> Dict:
+    if field.default_factory is not dataclasses.MISSING:
+        raise TypeError(field)
+    params = {
+        "required": field.default is dataclasses.MISSING,
+    }
+    if field.type is bool and not params["required"] and not field.default:
+        params.update({"action": "store_true"})
+        return params
+    if not params["required"]:
+        params.update({"default": field.default})
+    params.update({
+        "action": "store",
+        "type": field.type,
+    })
+    return params
+
+
+def _add_dataclass_arguments(parser: argparse.ArgumentParser, cls: type):
+    assert dataclasses.is_dataclass(cls)
+    for field in dataclasses.fields(cls):
+        params = _field_to_argument_params(field)
+        parser.add_argument(f"--{field.name}", **params)
+
+
+def main(prog_args = None) -> int:
+    parser = argparse.ArgumentParser(
+        "Compile a (MLIR) .mlir file to verilog (.v/.sv)"
+    )
+
+    parser.add_argument(
+        "infile",
+        metavar="<input filename>",
+        action="store",
+        type=str,
+        help="Input MLIR file",
+    )
+    parser.add_argument(
+        "--outfile",
+        metavar="<output filename>",
+        action="store",
+        type=argparse.FileType("w"),
+        required=False,
+        default=sys.stdout,
+    )
+    _add_dataclass_arguments(parser, MlirToVerilogOpts)
+
+    args = parser.parse_args(prog_args)
+    opts = slice_opts(vars(args), MlirToVerilogOpts)
+    
+    logging.debug(f"Running with opts: {opts}")
+    if opts.split_verilog and args.outfile is not sys.stdout:
+        logging.warning(
+            f"outfile ({args.outfile.name}) likely ignored with split_verilog "
+            f"enabled"
+        )
+
+    with open(args.infile, "r") as f_in:
+        mlir_to_verilog(f_in, args.outfile, opts)
+
+
+if __name__ == "__main__":
+    exit(main())
+    
```

Comment:
Let's add a test to cover this script

Reply:
Is it worth adding this script to the package (like https://github.com/phanrahan/magma/blob/master/setup.py#L15) so that it's easier to use? That way we wouldn't have to manually add it to the path or find it relative to current location.  We could prefix with `magma_mlir_to_verilog` or something so it doesn't conflict in the path.  Also, perhaps there's a way to only install it in dev mode unless we think the user might find it useful.  I know there's `python setup.py develop`, not sure if there's a pip equivalent.

Reply:
That's a good idea. So I've intended everything in `tools/` to be run as `python tools/{tool}.py`. I'm ok leaving that as is since it's a developer only thing (develop setup mode nonetheless noted) and it's just the filepath and they're all in tools.